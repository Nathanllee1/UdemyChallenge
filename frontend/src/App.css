* {
  background-color: #E9B44C;
  color: #1C110A;

}

.content {
  display: flex;
  justify-content: center;
  align-items: center;
}

.intro {
  text-align: center;
}

h1 {

}

button {
  width: 100px;
  height: 45px;
  background-color: #50A2A7;
}
