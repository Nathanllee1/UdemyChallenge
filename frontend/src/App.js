import React from 'react';
import './App.css';

import Question from './question';
import Results from './results';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      questions : [],
      questionCount: 0,
      currentScreen : 'Questions',
      answers: [],
      answerObject: []
    };
  };

  componentDidMount() {
    fetch("/questions")
      .then((response) => {
        return response.json()
      })
      .then(data => this.setState({questions : data}))
      .catch((error) => {
        console.log("Couldn't fetch data")
      });
  };

  onQuestionSubmit = () => {
    this.setState({currentScreen : 'Results'});
    fetch('/answers', {
      method: 'POST', // or 'PUT'
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(this.state.answers),
    })
    .then((response) => {
      return response.json()
    })
    .then((data) => {
      this.setState({answerObject : data});
    })
    .catch((error) => {
      console.log('Could not submit data');
    });
  };

  onQuestionChange = (e) => {
    var currentQuestions = this.state.answers;

    currentQuestions[[e.target.name]] = e.target.value;

    this.setState({ answers : currentQuestions });
  };

  render() {
    if (this.state.currentScreen === 'Questions') {
      return (
        <div className="content">
          <div className="intro">
            <div className="introBox">
              <h1>Welcome Udemy!</h1>
              <h2>Here's a quiz about me, <b>Nathan Lee</b> </h2>
              <p>(Hint: my resume might help)</p><br/>
              <h3>My response to the <a href="https://github.com/udemy/coding-challenge">Udemy Web App Coding Challenge</a></h3>
              <h4>The <a href="https://github.com/Nathanllee1/UdemyChallenge">code and writeup</a></h4> <br/>
              <a type="button" className="btn btn-primary" href="#q0">Begin Quiz</a>
            </div>
          </div>
          <form onSubmit={this.onQuestionSubmit} onChange={this.onQuestionChange}>
            {this.state.questions.map((question, index) => (
              <Question key={index} question={question} questionNumber={index} onSubmit={this.onQuestionSubmit} lastQuestion={this.state.questions.length}/>
            ))}
          </form>
        </div>
      );
    } else if (this.state.currentScreen === 'Results') {
      return (
        <Results answerObject={this.state.answerObject} questions={this.state.questions}/>
      );
    };
  };
};

export default App;
