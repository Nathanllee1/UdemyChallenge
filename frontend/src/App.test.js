import React from 'react';
import App from './App';

import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

//Import components to test
import Question from './question';
import Results from './results';


//Setup
let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


it("Renders question series correctly after fetch", async () => {
  const fakeQuestions = [{"question" : "Why", "answers" : ['1', '2', '3']}, {"question": "Where", "answers" : ['1', '2', '3']}]

  jest.spyOn(global, "fetch").mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeQuestions)
    })
  )

  await act(async () => {
    render(<App />, container)
  })
  expect(container).toMatchSnapshot()
}),

it("Check results render the right results", () => {
  act(() => {
    render(<Results answerObject={[1, 1], [0, 1]} questions={[{"question":"?1", "answers":["Yes", "No"]}, {"question":"?2", "answers":["Perhaps", "Maybe"]}]}/>, container);
  });

  expect(container).toMatchSnapshot()
}),

it("Check question object displaying right amount of questions", () => {
  act(() => {
    render(<Question question={{"question":"?1", "answers":["Yes", "No"]}} questionNumber={1} />, container)
  });
  expect(container).toMatchSnapshot()
})
