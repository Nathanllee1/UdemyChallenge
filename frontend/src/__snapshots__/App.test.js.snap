// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check question object displaying right amount of questions 1`] = `
<div>
  <div
    class="question"
    id="q1"
  >
    <div
      class="questionBox"
    >
      <h1
        class="questionTitle"
      >
        ?1
      </h1>
      <div
        class="radio"
      >
        <label>
          <input
            name="1"
            type="radio"
            value="0"
          />
          Yes
        </label>
      </div>
      <div
        class="radio"
      >
        <label>
          <input
            name="1"
            type="radio"
            value="1"
          />
          No
        </label>
      </div>
      <input
        class="btn btn-primary"
        disabled=""
        type="submit"
        value="Submit Quiz"
      />
    </div>
  </div>
</div>
`;

exports[`Check results render the right results 1`] = `
<div>
  <div
    class="results"
  >
    <div
      class="resultsBox"
    >
      <h1>
        You got 
        2
         out of 
        2
         questions right
      </h1>
      <br />
      <div
        class="answerBox"
      >
        <h3>
          <b>
            Question 
            0
          </b>
          : 
          ?1
        </h3>
        <div
          class="correctAnswer"
        >
          <h4>
              
            <i
              class="fas fa-check"
            />
          </h4>
        </div>
        <br />
      </div>
      <div
        class="answerBox"
      >
        <h3>
          <b>
            Question 
            1
          </b>
          : 
          ?2
        </h3>
        <div
          class="correctAnswer"
        >
          <h4>
              
            <i
              class="fas fa-check"
            />
          </h4>
        </div>
        <br />
      </div>
      <a
        class="btn btn-primary"
        href="/"
        type="button"
      >
        Take Quiz Again
      </a>
    </div>
  </div>
</div>
`;

exports[`Renders question series correctly after fetch 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="intro"
    >
      <div
        class="introBox"
      >
        <h1>
          Welcome Udemy!
        </h1>
        <h2>
          Here's a quiz about me, 
          <b>
            Nathan Lee
          </b>
           
        </h2>
        <p>
          (Hint: my resume might help)
        </p>
        <br />
        <h3>
          My response to the 
          <a
            href="https://github.com/udemy/coding-challenge"
          >
            Udemy Web App Coding Challenge
          </a>
        </h3>
        <h4>
          The 
          <a
            href="https://github.com/Nathanllee1/UdemyChallenge"
          >
            code and writeup
          </a>
        </h4>
         
        <br />
        <a
          class="btn btn-primary"
          href="#q0"
          type="button"
        >
          Begin Quiz
        </a>
      </div>
    </div>
    <form>
      <div
        class="question"
        id="q0"
      >
        <div
          class="questionBox"
        >
          <h1
            class="questionTitle"
          >
            Why
          </h1>
          <div
            class="radio"
          >
            <label>
              <input
                name="0"
                type="radio"
                value="0"
              />
              1
            </label>
          </div>
          <div
            class="radio"
          >
            <label>
              <input
                name="0"
                type="radio"
                value="1"
              />
              2
            </label>
          </div>
          <div
            class="radio"
          >
            <label>
              <input
                name="0"
                type="radio"
                value="2"
              />
              3
            </label>
          </div>
          <a
            class="btn btn-primary disabled"
            href="#q1"
          >
            Next
          </a>
        </div>
      </div>
      <div
        class="question"
        id="q1"
      >
        <div
          class="questionBox"
        >
          <h1
            class="questionTitle"
          >
            Where
          </h1>
          <div
            class="radio"
          >
            <label>
              <input
                name="1"
                type="radio"
                value="0"
              />
              1
            </label>
          </div>
          <div
            class="radio"
          >
            <label>
              <input
                name="1"
                type="radio"
                value="1"
              />
              2
            </label>
          </div>
          <div
            class="radio"
          >
            <label>
              <input
                name="1"
                type="radio"
                value="2"
              />
              3
            </label>
          </div>
          <input
            class="btn btn-primary"
            disabled=""
            type="submit"
            value="Submit Quiz"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
